File: app/src/main/java/com/doubean/ford/ui/groups/groupsHome/GroupsHomeViewModel.java
package com.doubean.ford.ui.groups.groupsHome;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.Transformations;
import androidx.lifecycle.ViewModel;

import com.doubean.ford.data.repository.GroupRepository;
import com.doubean.ford.data.repository.GroupsFollowsAndSavesRepository;
import com.doubean.ford.data.vo.GroupDetail;
import com.doubean.ford.data.vo.GroupFollowItem;
import com.doubean.ford.data.vo.Resource;
import com.doubean.ford.util.LiveDataUtil;

import java.util.List;

public class GroupsHomeViewModel extends ViewModel {
    private final GroupRepository groupRepository;
    private final GroupsFollowsAndSavesRepository groupsFollowsAndSavesRepository;
    private final LiveData<List<GroupFollowItem>> followList;


    public GroupsHomeViewModel(@NonNull GroupRepository groupRepository, @NonNull GroupsFollowsAndSavesRepository groupsFollowsAndSavesRepository) {
        super();
        this.groupsFollowsAndSavesRepository = groupsFollowsAndSavesRepository;
        this.followList = createFollowListLiveData();
        this.groupRepository = groupRepository;

    }

    public LiveData<List<GroupFollowItem>> getFollowList() {
        return followList;
    }

    private LiveData<List<GroupFollowItem>> createFollowListLiveData() {
        return Transformations.switchMap(groupsFollowsAndSavesRepository.getFollowIds(), follows -> {
            LiveData<GroupFollowItem>[] array = new LiveData[follows.size()];
            for (int i = 0; i < array.length; i++) {
                LiveData<Resource<GroupDetail>> group = groupRepository.getGroup(follows.get(i).groupId, false);
                int finalI = i;
                array[i] = Transformations.map(group, g -> {
                    if (g != null && g.data != null) {
                        return new GroupFollowItem(follows.get(finalI).groupId, follows.get(finalI).groupTabId, g.data);
                    }
                    return null;
                });
            }
            return LiveDataUtil.zip(array);
        });

    }


}

File: app/src/main/java/com/doubean/ford/ui/groups/groupsHome/GroupsHomeViewModel.kt
package com.doubean.ford.ui.groups.groupsHome

import androidx.lifecycle.*
import com.doubean.ford.data.repository.GroupRepository
import com.doubean.ford.data.repository.GroupUserDataRepository
import com.doubean.ford.model.GroupRecommendationType
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.flowOn

class GroupsHomeViewModel(
    groupRepository: GroupRepository,
    groupUserDataRepository: GroupUserDataRepository,
) : ViewModel() {

    val follows = groupUserDataRepository.getAllGroupFollows().flowOn(Dispatchers.IO).asLiveData()

    val groupsOfTheDay =
        groupRepository.getGroupRecommendation(GroupRecommendationType.DAILY).flowOn(Dispatchers.IO)
            .asLiveData()

    companion object {
        class Factory(
            private val repository: GroupRepository,
            private val groupUserDataRepository: GroupUserDataRepository,
        ) : ViewModelProvider.NewInstanceFactory() {
            @Suppress("UNCHECKED_CAST")
            override fun <T : ViewModel> create(modelClass: Class<T>): T {
                return GroupsHomeViewModel(repository, groupUserDataRepository) as T
            }
        }
    }
}